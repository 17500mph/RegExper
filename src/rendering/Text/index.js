import React from 'react';
import PropTypes from 'prop-types';

import * as style from './style';

class Text extends React.PureComponent {
  static propTypes = {
    quoted: PropTypes.bool,
    transform: PropTypes.string,
    onReflow: PropTypes.func,
    children: PropTypes.oneOfType([
      PropTypes.arrayOf(PropTypes.node),
      PropTypes.node
    ]).isRequired
  }

  renderContent() {
    const { children, quoted } = this.props;

    if (!quoted) {
      return children;
    }

    return <>
      <tspan style={ style.quote }>&ldquo;</tspan>
      <tspan>{ children }</tspan>
      <tspan style={ style.quote }>&rdquo;</tspan>
    </>;
  }

  render() {
    const { transform } = this.props;

    const textProps = {
      style: style.text,
      transform
    };

    return <text { ...textProps }>
      { this.renderContent() }
    </text>;
  }
}

export default Text;
