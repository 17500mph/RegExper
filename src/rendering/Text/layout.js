import React from 'react';
import ReactDOM from 'react-dom';

import Text from 'rendering/Text';

const layoutText = data => {
  const container = document.createElement('div');
  document.body.appendChild(container);

  ReactDOM.render(
    <svg width="0" height="0" viewBox="0 0 0 0">
      <Text { ...data.props }>{ data.children }</Text>
    </svg>,
    container);

  const box = container.querySelector('svg > text').getBBox();
  document.body.removeChild(container);

  data.box = {
    width: box.width,
    height: box.height
  };
  data.props.transform = `translate(${ -box.x } ${ -box.y })`;
  return data;
};

export default layoutText;
