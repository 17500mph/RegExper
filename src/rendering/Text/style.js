import { fontSize, fontFamily, grey } from 'rendering/style';

export const text = {
  fontSize, fontFamily
};
export const quote = {
  fill: grey
};
