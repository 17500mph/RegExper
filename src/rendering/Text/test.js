import React from 'react';
import { render } from 'react-testing-library';

import Text, { layout } from 'rendering/Text';

describe('Text', () => {
  test('rendering', () => {
    const { asFragment } = render(
      <svg>
        <Text>Example</Text>
      </svg>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  test('positioning text', () => {
    const { asFragment } = render(
      <svg>
        <Text transform="translate(1 2)">Example</Text>
      </svg>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  test('rendering with a theme', () => {
    const { asFragment } = render(
      <svg>
        <Text theme="anchor">Example</Text>
      </svg>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  test('rendering quoted text', () => {
    const { asFragment } = render(
      <svg>
        <Text quoted={ true }>Example</Text>
      </svg>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  test('layout', () => {
    global.Element.prototype.getBBox.mockReturnValue({
      width: 100,
      height: 20,
      y: 18
    });

    const processed = layout({
      type: 'Text',
      props: {
        property: 'example'
      },
      children: [
        'Example content'
      ]
    });
    expect(processed).toMatchSnapshot();
  });
});
