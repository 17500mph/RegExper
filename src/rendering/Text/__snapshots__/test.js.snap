// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Text layout 1`] = `
Object {
  "box": Object {
    "height": 20,
    "width": 100,
  },
  "children": Array [
    "Example content",
  ],
  "props": Object {
    "property": "example",
    "transform": "translate(0 -18)",
  },
  "type": "Text",
}
`;

exports[`Text positioning text 1`] = `
<DocumentFragment>
  <svg>
    <text
      style="font-size: 16px; font-family: Arial;"
      transform="translate(1 2)"
    >
      Example
    </text>
  </svg>
</DocumentFragment>
`;

exports[`Text rendering 1`] = `
<DocumentFragment>
  <svg>
    <text
      style="font-size: 16px; font-family: Arial;"
    >
      Example
    </text>
  </svg>
</DocumentFragment>
`;

exports[`Text rendering quoted text 1`] = `
<DocumentFragment>
  <svg>
    <text
      style="font-size: 16px; font-family: Arial;"
    >
      <tspan
        style="fill: #908c83;"
      >
        “
      </tspan>
      <tspan>
        Example
      </tspan>
      <tspan
        style="fill: #908c83;"
      >
        ”
      </tspan>
    </text>
  </svg>
</DocumentFragment>
`;

exports[`Text rendering with a theme 1`] = `
<DocumentFragment>
  <svg>
    <text
      style="font-size: 16px; font-family: Arial; fill: #fff;"
    >
      Example
    </text>
  </svg>
</DocumentFragment>
`;
